### PR TITLE
Bump kubernetes to 1.16.0

### PR DESCRIPTION
**Pull request by mfojtik**
For review:

- [x] @deads2k 1f6898b5 INTERESTING: remove GetInitializers/SetInitializers
- [x] @soltysh d9e7e123 BORING: adjust TablePrinter at REST
- [x] @soltysh 4c7f2f59 INTERESTING: refactor printers to table printers
- [x] @sttts 8c4d506e BORING: fix SchemeBasedObjectInterfaces
- [ ] @sttts 75af5409 INTERESTING: adjust Delete REST – follow-up
- [x] @deads2k 92d587b1 INTERESTING: hack/openapi-violation.list changed
- [x] @sttts b083e5a9 BORING: update generated files
- [x] @enj 751d3e30 INTERESTING: move rbac and core clients to external
- [x] @brancz @deads2k 8fed2099 INTERESTING: remove reflector prometheus metrics
- [x] @sttts a70d80d4 BORING: move k8s.io/kubernetes/pkg/api/ref
- [x] @sttts 7f583010 BORING: move k8s.io/apiserver/pkg/storage/etcd/testing
- [x] @soltysh d3d72f81 BORING: move kubectl utils and templates

### CODE REVIEW

File: glide.yaml [hunk 1]
```diff
@@ -2,85 +2,84 @@ package: github.com/openshift/openshift-apiserver
...
-  version: kubernetes-1.14.0
+  version: kubernetes-1.16.0
 - package: k8s.io/sample-controller
-  version: kubernetes-1.14.0
+  version: kubernetes-1.16.0
 - package: k8s.io/kubernetes
-  repo:    https://github.com/openshift/kubernetes.git
-  version: openshift-apiserver-4.2-kubernetes-1.14.0
- # this matches the 1.14 branch from kube
+  version: v1.16.0-beta.2
```
Comment (soltysh) [inline glide.yaml]: @mfojtik I think you missed this one

File: pkg/apps/apiserver/registry/instantiate/rest.go [hunk 2]
```diff
@@ -106,7 +106,8 @@ func (s *REST) Create(ctx context.Context, obj runtime.Object, createValidation
 		config.Status.LatestVersion++
 
 		userInfo, _ := apirequest.UserFrom(ctx)
-		attrs := admission.NewAttributesRecord(config, old, apps.Kind("DeploymentConfig").WithVersion(""), config.Namespace, config.Name, apps.Resource("DeploymentConfig").WithVersion(""), "", admission.Update, false, userInfo)
+		attrs := admission.NewAttributesRecord(config, old, apps.Kind("DeploymentConfig").WithVersion(""), config.Namespace, config.Name, apps.Resource("DeploymentConfig").WithVersion(""), "", admission.Update,
```
Comment (deads2k) [inline pkg/apps/apiserver/registry/instantiate/rest.go]: pass the options

**Comment by deads2k** [comment]
`@sttts 75af540 INTERESTING: adjust Delete REST`, you need to pass the options through.

**Comment by mfojtik** [comment]
https://github.com/openshift/openshift-apiserver/commit/75af540908a4fde7da91cee569df075baf7d23f0#r35209365

this will be addressed as follow up, the blocker BZ will be created for 4.3.

File: glide.yaml [hunk 3]
```diff
@@ -2,75 +2,74 @@ package: github.com/openshift/openshift-apiserver
...
 - package: github.com/coreos/bbolt
   version: v1.3.1-coreos.6
 - package: github.com/google/cadvisor
-  version: v0.32.0
-# for gengo
-- package: golang.org/x/tools/imports
-  version: 2382e3994d48b1d22acc2c86bcad0a2aff028e32
+  version: 9fa3b1429966 # kube says v0.33.2 but that tag does not exists
+- package: sigs.k8s.io/structured-merge-diff
+  version: 0c1d754dd648 # matches 1.16.0-beta.1 go.mod
```
Comment (sttts) [inline glide.yaml]: this certainly has changed as SSA merged later.

File: pkg/apps/apis/apps/v1/zz_generated.defaults.go [hunk 4]
```diff
@@ -195,9 +201,6 @@ func SetObjectDefaults_DeploymentConfig(in *v1.DeploymentConfig) {
 					}
 				}
 			}
-			if a.SecurityContext != nil {
-				corev1.SetDefaults_SecurityContext(a.SecurityContext)
```
Comment (sttts) [inline pkg/apps/apis/apps/v1/zz_generated.defaults.go]: why is this gone?

File: pkg/image/apiserver/registry/imagestream/registry.go [hunk 5]
```diff
@@ -105,7 +105,7 @@ func (s *storage) UpdateImageStreamStatus(ctx context.Context, imageStream *imag
 }
 
 func (s *storage) DeleteImageStream(ctx context.Context, imageStreamID string) (*metav1.Status, error) {
-	obj, _, err := s.Delete(ctx, imageStreamID, nil)
+	obj, _, err := s.Delete(ctx, imageStreamID, nil, nil)
```
Comment (sttts) [inline pkg/image/apiserver/registry/imagestream/registry.go]: rest.ValidateAllObjectFunc ?

File: pkg/authorization/apiserver/registry/clusterrole/proxy.go [hunk 6]
```diff
@@ -93,7 +93,7 @@ func (s *REST) Get(ctx context.Context, name string, options *metav1.GetOptions)
 	return role, nil
 }
 
-func (s *REST) Delete(ctx context.Context, name string, options *metav1.DeleteOptions) (runtime.Object, bool, error) {
+func (s *REST) Delete(ctx context.Context, name string, objectFunc rest.ValidateObjectFunc, options *metav1.DeleteOptions) (runtime.Object, bool, error) {
 	client, err := s.getImpersonatingClient(ctx)
```
Comment (sttts) [inline pkg/authorization/apiserver/registry/clusterrole/proxy.go]: `// TODO: call objectFunc`, and everywhere below. 

what is the BZ to fix those?

File: pkg/project/util/util.go [hunk 7]
```diff
@@ -31,6 +34,18 @@ func ConvertNamespaceFromExternal(namespace *corev1.Namespace) *projectapi.Proje
 			Phase: kapi.NamespacePhase(namespace.Status.Phase),
 		},
 	}
+	if namespace.Status.Conditions != nil {
+		p.Status.Conditions = []kapi.NamespaceCondition{}
+		status := &kapi.NamespaceStatus{}
+		if err := v1.Convert_v1_NamespaceStatus_To_core_NamespaceStatus(&namespace.Status, status, nil); err != nil {
+			panic(err)
```
Comment (sttts) [inline pkg/project/util/util.go]: handle error correctly

File: pkg/project/util/util.go [hunk 8]
```diff
@@ -31,6 +34,18 @@ func ConvertNamespaceFromExternal(namespace *corev1.Namespace) *projectapi.Proje
...
+		}
+		for _, c := range status.Conditions {
```
Comment (sttts) [inline pkg/project/util/util.go]: why the loop?

File: pkg/project/util/util.go [hunk 9]
```diff
@@ -48,6 +63,16 @@ func ConvertProjectToExternal(project *projectapi.Project) *corev1.Namespace {
 			Phase: corev1.NamespacePhase(project.Status.Phase),
 		},
 	}
+	if project.Status.Conditions != nil {
+		namespace.Status.Conditions = []corev1.NamespaceCondition{}
+		status := &projectv1.ProjectStatus{}
+		if err := projectinternalv1.Convert_project_ProjectStatus_To_v1_ProjectStatus(&project.Status, status, nil); err != nil {
+			panic(err)
```
Comment (sttts) [inline pkg/project/util/util.go]: handle error

File: pkg/project/util/util.go [hunk 10]
```diff
@@ -48,6 +63,16 @@ func ConvertProjectToExternal(project *projectapi.Project) *corev1.Namespace {
...
+		}
+		for _, c := range status.Conditions {
```
Comment (sttts) [inline pkg/project/util/util.go]: why the loop?

File: pkg/image/apiserver/registry/imagestream/registry.go (hunk 5 above)
Comment (mfojtik) [reply pkg/image/apiserver/registry/imagestream/registry.go]:
```
// ValidateAllObjectFunc is a "admit everything" instance of ValidateObjectFunc.
func ValidateAllObjectFunc(ctx context.Context, obj runtime.Object) error {
	return nil
}
```

but yes.

File: pkg/authorization/apiserver/registry/clusterrole/proxy.go (hunk 6 above)
Comment (mfojtik) [reply pkg/authorization/apiserver/registry/clusterrole/proxy.go]: https://bugzilla.redhat.com/show_bug.cgi?id=1754918

File: pkg/project/util/util.go (hunk 7 above)
Comment (mfojtik) [reply pkg/project/util/util.go]: @sttts too many changes, fine with follow up?

File: pkg/project/util/util.go (hunk 8 above)
Comment (mfojtik) [reply pkg/project/util/util.go]: why not?

File: pkg/apps/apis/apps/v1/zz_generated.defaults.go (hunk 4 above)
Comment (mfojtik) [reply pkg/apps/apis/apps/v1/zz_generated.defaults.go]: @tnozicka @soltysh FYI

Comment (tnozicka) [reply pkg/apps/apis/apps/v1/zz_generated.defaults.go]: it was before generated with f40157d9638d05ebf65421a9bbad5a7f68878b65 from https://github.com/openshift/kubernetes-code-generator.git

https://github.com/openshift/origin/blob/d4afc455d4b77b384f49dea1629a66236d213bba/glide.lock#L2094-L2097

Comment (mfojtik) [reply pkg/apps/apis/apps/v1/zz_generated.defaults.go]: this change seems fine as the `SetDefaults_SecurityContext` does not exits upstream and the SecurityContext seems to be defaulted inside Pod defaulter (set to empty struct) ?

Comment (sttts) [reply pkg/apps/apis/apps/v1/zz_generated.defaults.go]: Was removed in https://github.com/kubernetes/kubernetes/pull/78885.

File: pkg/authorization/apiserver/registry/clusterrole/proxy.go (hunk 6 above)
Comment (sttts) [reply pkg/authorization/apiserver/registry/clusterrole/proxy.go]: What's the plan now? Document? Already done in openshift/api?

Comment (mfojtik) [reply pkg/authorization/apiserver/registry/clusterrole/proxy.go]: @sttts i updated the BZ

**Pull request merged**